docs/guide/api-vite-environment.md

```diff
@@ -0,0 +1,673 @@
+:::
+
+A single Vite dev server can be used to interact with different module execution environments concurrently. We'll use the word environment to refer to a configured Vite processing pipeline that can resolve ids, load, and process source code and is connected to a runtime where the code is executed. Some examples of environments are browser, ssr, workerd, rsc. The transformed source code is called a module, and the relationships between the modules processed in each environment are kept in a module graph. The code for these modules is sent to the runtimes associated with each environment to be executed. When a module is evaluated, the runtime will request its imported modules triggering the processing of a section of the module graph.
+
+A Vite Module Runner allows running any code by processing it with Vite plugins first. It is different from `server.ssrLoadModule` because the runtime implementation is decoupled from the server. This allows library and framework authors to implement their layer of communication between the Vite server and the runner. The browser communicates with its corresponding environment using the server Web Socket and through HTTP requests. The SSR Module runner can directly do function calls to process modules. Other environments could run modules connecting to an edge runtime like workerd, or a Worker Thread in the same node process as Vitest does.
+
+All these environments share Vite's HTTP server, middlewares, and Web Socket. The resolved config and plugins pipeline are also shared, but plugins can use `apply` so its hooks are only called for certain environments. The environment can also be accessed inside hooks for fine-grained control.
+
+![Vite Environments](../images/vite-environments.svg)
+
+## Using environments in the Vite server
+
+A Vite dev server exposes two environments by default named `'browser'` and `'ssr'`. The browser environment runs in client apps that have imported the `/@vite/client` module. The SSR environment runs in the same runtime as the Vite server (for example, in node) and allows application servers to be used to render requests during dev with full HMR support. We'll discuss later how frameworks and users can create and register new environments.
+
+The available environments can be accessed using the `server.environments` read-only array:
+
+```js
+server.environments.forEach((environment) => log(environment.name))
+```
+
+Each environment can also be accessed through its name:
+
+```js
+server.environment('browser').transformRequest(url)
+```
```

Comment:
I think the shape of `server.environments` is confusing. It looks like a function and an array at the same time. I think we need to decide if it should be an array, Set, or an object shape. Maybe `server.environment('browser')` can be `server.environments.get('browser')`? It seems like the goal is to prevent mutations.

Reply:
A map would be great, but we wanted an array because we need to have an order to call `hotUpdate` for example, and also to build the environments if we decide to get into that game.

But we could have that order in a different way. We would need these to be in sync but maybe just an array of names is ok here, also for configuration (`server.hmr.environments: ['ssr', 'browser']`, `build.environments: ['browser', 'ssr']`). I think some folks wanted SSR first for HMR, but most frameworks build the client first.

It would be fine for me to switch to a map and `server.environments.get('browser')`. But even then, I see value in offering `server.environment('browser')` as a shortcut to whatever we choose as it reads better to me.

Reply:
Just an idea, not convinced this is good, but it may be interesting. We could have a function define the order instead of using arrays so users can decide to execute in parallel too if their frameworks are fine with that:
```js
// Access
server.environment('name')

// Adding new environments using a plugin hook
registerEnvironments(server, environments: Set<Environment>) {
  // ...
  environments.add(workerdEnvironment)
}

// Define order for `hotUpdate` using a config option
server.modulesChange = (server, handleHotUpdate: (environment: string) => Promise<void> {
  // Orchestrate handling of hot updates in series
  await handleHotUpdate('ssr')
  await handleHotUpdate('browser')

  // or do it in parallel
}

// Define order for build time using a config option
build.defineBuilds = (builder, build: (environment: string) => Promise<void>) {
  // Orchestrate builds
  await build('browser')
  await build('ssr')

  // or do it in parallel
}
```
I think we would still need a way to query all environments, as in `server.environments`. Maybe we should have environments be a Map<string, environment> as proposed before.

---

Something else about a Map/Set vs Array that we discussed with Vladimir. Using an array and trying to avoid the use of names as a unique identifier, would potentially allow having two environments of the same type. Vlad was thinking about Vitest monorepo support and how he could use a JSDom environment for each package. It isn't clear to me or him if this is really needed. Just sharing as we could try only using instances to avoid referring to them by name. Or maybe we need an environment `id` and a `name`. So the id is unique `{package-name}:jsdom` and the name is not `jsdom`.

Reply:
I'm not sure how ordered environments will work yet, as you mentioned different frameworks have different needs. The ones I'm aware of are:

- SvelteKit: Runs `server`, `client`, then `worker` (service worker, optional) builds.
- Astro: Runs `server` then `client` builds.

But even so, I don't think we should make the order of environments matter. All environments tasks like `hotUpdate` should kick off at once. It would be weird if the `ssr` HMR have to wait for the `browser` HMR to finish for example. We also wouldn't want Vite plugins to have make assumptions of the order too.

I think we should tackle the environment builds concept later, but the idea of annotating the "type" of build this way is interesting.

---

Re `server.environments`, it's not clear to me what we want to expose according to the docs. I'm not sure I can suggest something correct without missing a goal you have 😅 I was thinking we're exposing a plain object that only has the `forEach` and `get` keys.

Reply:
I'm liking more and more the idea of using a config function to orchestrate builds and HMR propagation. It is easy to maintain and implement for us, and would let frameworks do whatever they want (series vs parallel, or a mix of them, and add logic between builds if needed too without relying on `buildEnd`).

About HMR propagation order for environments, we got a request for this feature here: https://github.com/vitejs/vite/discussions/15774#discussioncomment-8575835. I also coded it as parallel as you suggest in the separate module graphs PR. We can have it in parallel by default and only add a config option function to orchestrate if HMR should be in series or parallel, or a mix, when we have good use cases (if the request I linked is not enough for that already).

---

I think the requirements for environments are:
- Quickly get an environment given its name (or id if we go there)
- Get a list of all environments (or expose forEach) so plugins can check if an environment is there already and add it if not. And to do things like logging or any operation on all environments when you don't know exactly what environments are configured.
- Register an environment. Done in a hook in the proposal, outside of the data structure.

Reply:
> It would be weird if the ssr HMR have to wait for the browser HMR to finish for example.

Can you explain why this is weird? Folks from Vike explained that then need to resolve ssr hmr first. Right now they are already doing this by calling it inside handleHotUpdate before returning new modules. If we can provide a built-in solution, that would be great

Reply:
It's weird because two environments may have delayed HMR feedback if one relies on the other. It feels brittle if HMR has to rely on a specific order of an environment.

Reply:
@nitedani would you explain why you need to wait for HMR propagation to be done in SSR before starting with the browser propagation?

Reply:
[I'm restarting the server runtime child process when the server code changes.](https://github.com/vikejs/vike/pull/1513/files#diff-e793b1e19f4b3aaf303c35365faf66c66bf3bd9da518311fb2a991eea715f3a7R51) If the browser reload doesn't wait for the server restart, the server might exit/restart during a request/response, causing an error page in the browser. If the browser is reloaded only after the server restart, it stays in loading state until the [server is listening](https://github.com/vikejs/vike/pull/1513/files#diff-6b65b38849e385354eaa6b29d3bca0954a5923549eba16a9415be870b1e1fbcdR28).

Reply:
Remix would _love_ if Vite had first class support for ordering/orchestrating builds. We wrote a bit about why [Remix couldn't just be a Vite plugin](https://remix.run/blog/remix-vite-stable#more-than-just-a-vite-plugin) anymore.

Here's some relevant code where we run the [client build first and then run N server builds](https://github.com/remix-run/remix/blob/main/packages/remix-dev/vite/build.ts#L288-L316) for our server bundles feature (N defaults to 1).

Reply:
For reference, there are many frameworks in the same situation and there has been several attempts to move orchestration of builds into core in the past:
- https://github.com/vitejs/vite/issues/5936
- https://github.com/vitejs/vite/discussions/9496

We discussed this quite a lot in the team, and the last consensus was that Vite didn't want to get into that game. See [comment from Anthony](https://github.com/vitejs/vite/discussions/9496#discussioncomment-3384709) summarising the position of the team at that point.

This didn't prevented most of the frameworks to hack their way into being a plugin because the benefits outweighed the hacks for them (like starting a new vite build at in the `writeBundle` hook of the prev build). SvelteKit, SolidStart, Remix, Qwik City, Vike, and a lot of others are in this same situation.

Looking back, I think it was maybe a good idea to avoid making this request a core feature at that point. We didn't have the concepts in place to go there first. I think that if we have a proper Environment high-level concept now, the position of several team members will be different and there is a better chance for `vite build` to take more responsabilities.

Reply:
Will Rolldown transpile the server- and client-side at the same time?

This question is one reason why I'm inclined to think that it's best if Vite owns the sequencing of the build. In other words, that it should be Vite's responsibility to determine what happens when `$ vite build` is run while introducing new options such as `build.parallel: boolean`.

Another small rationale is [`ssrEmitAssets`](https://github.com/vitejs/vite/pull/11430) which every SSR framework needs to use and implement. If Vite owns the build sequencing then Vite could implement that on behalf of SSR frameworks.

<hr>

That said, in order to enable plugins such `vite-plugin-pwa`, `@vitejs/plugin-legacy` and the potentially upcoming hosting Vite plugins (Cloudlare/Netlify/Vercel/...), it may make sense if Vite provides an official way to hook into the build process. I guess whether this is needed depends on how the Vite Environment API evolves.

<hr>

@pcattori I can see the marketing appeal of it, but I recommend Remix against doing that. In my experience (Vike aka vite-plugin-ssr was amongst the first, if not the first, framework to do that), (mis-)using the Vite CLI is causing issues for zero technical gain. The only gain I can see is a marketing one. [Quoting antfu](https://github.com/vitejs/vite/discussions/9496#discussioncomment-3384709):

> We recommend higher-level integrations to provide their own CLI to fit various needs. For example, a framework could proxy other Vite subcommands but only provide custom logic for its build command.

I agree.

Simple examples:
 - Imagine you want to have a new CLI flag to disable pre-rendering: it's confusing for the user that `$ vite build --disable-pre-rendering` is a framework-specific CLI argument.
 - Imagine you want to introduce a new CLI command that lists all the app's routes. It's confusing for the user to use `$ vite routes`  that is framework specific, or having to use two CLIs `$ vite` and `$ remix`.

At the end of the day I think it boils down that: beyond the marketing aspect of it, is there any benefit of (mis-)using Vite's CLI?

Funny coincidence: using a dedicated Vike CLI instead of (mis-)using Vite's CLI is actually the next big task in Vike's pipeline.

Reply:
- Cannot (or hacky / weird for the user) extend new CLI commands & arguments. (See the two examples of my previous reply.)
- [Need to duplicate and maintain a copy of all Vite CLI options](https://github.com/vikejs/vike/blob/b65c343ddb6da0904f16970e0d8b0091830314b4/vike/node/plugin/plugins/autoFullBuild.ts#L75).
- Vite version mismatch. (Vite CLI resolves to the user-land `node_modules/vite/`, whereas the Vite API resolves to Vike-land `node_modules/.pnpm/vike/node_modules/vite/`).
- [Vike specific] Vike now directly uses esbuild to transpile Vike config files. Vike needs to resolves its config before spinning up the Vite instance. It's a hard requirement for the upcoming Vike versions, thus Vike cannot use the Vite CLI anymore.
- [Vike specific][Experimental] Vike is going to forgo the entire concept of CLI arguments and instead go for `$ VIKE={prerender:true,ssr:false} vike build`.